server/service.go

```diff
@@ -147,9 +157,9 @@ func (m *BoostService) RegisterValidatorV1(ctx context.Context, message types.Re
 
 	// If no relay responded true, return the last error message, or a generic error
 	var err error
-	if !ok {
-		err = lastRelayError
-		if lastRelayError == nil {
+	if !container.ok {
+		err = container.lastRelayError
+		if container.lastRelayError == nil {
```

Comment:
I think the `defer container.mu.Unlock()` at L150 will be executed on the completion of that goroutine... which means we probably want to hold the read lock here as well if we assume `RegisterValidatorV1` will be called concurrently

Reply:
Correct me if I'm wrong, but if `RegisterValidatorV1` is called by several goroutines in the JSON-RPC dispatcher, its internal values will not be shared across them ?

To simplify, I've written the following example :
```go
package main

import (
	"fmt"
	"sync"
)

func RegisterValidatorV1(wg *sync.WaitGroup, c chan bool) {
	defer wg.Done()

	type safeContainer struct {
		mu   sync.Mutex
		data bool
	}

	container := safeContainer{}

	var pool sync.WaitGroup

	for i := 0; i < 30; i++ {
		pool.Add(1)
		go func() {
			defer pool.Done()

			// Do some stuff, such as inverting container data value
			container.mu.Lock()
			defer container.mu.Unlock()

			container.data = !container.data
		}()
	}

	pool.Wait()

	<-c
	// Read container data
	fmt.Printf("container.data address: %p\n", &container.data)
}

func main() {
	var wg sync.WaitGroup
	defer wg.Wait()

	// We'll use this channel to start the goroutines at the same time
	c := make(chan bool)

	for i := 0; i < 50; i++ {
		wg.Add(1)
		go RegisterValidatorV1(&wg, c)
	}
	close(c)
}
```

Here, the `c` channel will ensure that the 50 initial goroutines will try to read `container.data` at the same time.
But as `container.data` is a local variable of the `RegisterValidatorV1`, I don't think there could be any data race here.

Reply:
> Correct me if I'm wrong, but if RegisterValidatorV1 is called by several goroutines in the JSON-RPC dispatcher, its internal values will not be shared across them ?

That is correct.

Reply:
That makes sense. Thanks for the explanations!